veps/sig-compute/hypervisor-abstraction.md

```diff
@@ -0,0 +1,481 @@
+
+## Motivation
+
+- KubeVirt currently hard-codes KVM/QEMU assumptions throughout virt-launcher, virt-handler, virt-controller, and node preparation scripts.
+- Platform teams that rely on alternative accelerators or hypervisor stacks face invasive forks to replace `/dev/kvm`, libvirt domain types, or resource scheduling hints.
+- A scoped abstraction keeps the project approachable for contributors while unlocking new hardware backends.
+
+## Goals
+
+- Document the cluster-wide hypervisor configuration and per-component extension points (defaults, converter, webhooks, node labeller) so downstream implementations can extend behavior without invasive changes to existing components.
+- Resolve the active hypervisor early and feed it through the virt-launcher converter so hypervisor-specific behavior stays localized.
+- Support both admission-time validation and mutation so administrators can enforce guardrails while still customizing VMIs for a given hypervisor.
+- Let components request allocatable device resources declared by the hypervisor configuration, avoiding new scheduling primitives.
+- Make it simple for downstreams to implement new hypervisors by following a documented contract.
+
+## Non Goals
+
+- Deliver a full implementation of any specific new hypervisor backend.
+- Redesign the VirtualMachineInstance API schema beyond additive fields.
+- Replace existing Hyper-V enlightenment features or other architecture-specific helpers.
+- Mandate new observability requirements; telemetry hooks remain optional.
+
+## Definition of Users
+
+- Cluster administrators who need to bootstrap KubeVirt on hardware that exposes alternative virtualization devices.
+- Platform vendors integrating proprietary or emerging hypervisor stacks with KubeVirt.
+- Upstream contributors maintaining virt-launcher, virt-controller, and virt-handler.
+
+## User Stories
+
+1. As a cluster administrator, I can declare a non-KVM hypervisor as the cluster default, and VMI pods schedule only on nodes that expose its required devices.
+2. As a platform engineer, I can supply hypervisor-specific VMI spec mutations and libvirt domain adjustments without forking the virt-launcher converter.
+3. As an upstream maintainer, I know exactly where to add validation, testing, and documentation when a new hypervisor is introduced.
+
+## Repos
+
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/enhancements](https://github.com/kubevirt/enhancements) (this VEP)
+
+## Design
+
+### Hypervisor Extension Points
+
+Cluster configuration (`spec.configuration.hypervisorConfiguration.name`) declares the active hypervisor for the entire installation, and each control-plane package exposes focused extension contracts so downstream implementations only touch the areas they actually need:
```

Comment:
IIUC this implies the hypervisor needs to be specified in the Kubevirt CR.
How about having another section that outlines what is the relation and separation proposed here.
There are multiple options on how to introduce new implementations, from compilation/build, deployment, through code conditions and up to interactions with optional components.

Decisions on having the implementation of new hypervisors embedded into the project codebase or as external extensions through well defined API interfaces is another important point.

Having this discussed in its own section may allow presenting the decision clearly and open an option to propose alternatives.

Reply:
If `spec.configuration.hypervisorConfiguration.name` is added, there is a risk that a future hypervisor expert would use it in an inappropriate place, complicating the core logic of kubevirt.
Let us assume that in two years time, one hypervisor realizes that they need to deploy a new component on relevant nodes. It would be very tempting for them to add an `if spec.configuration.hypervisorConfiguration.name=="bla"` into virt-operator, when the proper solution means introducing a plugin mechanism to virt-operator.

Reply:
@harshitgupta1337 thanks for changing the API to allow multiple hypervisors in the future.
However, would you add to the VEP a discussion about why we prefer to have it at all? I think that most cluster admin would prefer to avoid this configuration, letting kubevirt iterate over the worker nodes to find the most prevalent hypervisor. Even if we realize that having an explicit configuration is better, it is far easier to /add/ the field than to drop it.